<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork26802033-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo"  />
    </a>
</p>


# **Launch Sites Locations Analysis with Folium**


Estimated time needed: **40** minutes


The launch success rate may depend on many factors such as payload mass, orbit type, and so on. It may also depend on the location and proximities of a launch site, i.e., the initial position of rocket trajectories. Finding an optimal location for building a launch site certainly involves many factors and hopefully we could discover some of the factors by analyzing the existing launch site locations.


In the previous exploratory data analysis labs, you have visualized the SpaceX launch dataset using `matplotlib` and `seaborn` and discovered some preliminary correlations between the launch site and success rates. In this lab, you will be performing more interactive visual analytics using `Folium`.


## Objectives


This lab contains the following tasks:

*   **TASK 1:** Mark all launch sites on a map
*   **TASK 2:** Mark the success/failed launches for each site on the map
*   **TASK 3:** Calculate the distances between a launch site to its proximities

After completed the above tasks, you should be able to find some geographical patterns about launch sites.


Let's first import required Python packages for this lab:


In [1]:
import piplite
await piplite.install(['folium'])
await piplite.install(['pandas'])

In [2]:
import folium
import pandas as pd

In [3]:
# Import folium MarkerCluster plugin
from folium.plugins import MarkerCluster
# Import folium MousePosition plugin
from folium.plugins import MousePosition
# Import folium DivIcon plugin
from folium.features import DivIcon

If you need to refresh your memory about folium, you may download and refer to this previous folium lab:


[Generating Maps with Python](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module\_3/DV0101EN-3-5-1-Generating-Maps-in-Python-py-v2.0.ipynb)


## Task 1: Mark all launch sites on a map


First, let's try to add each site's location on a map using site's latitude and longitude coordinates


The following dataset with the name `spacex_launch_geo.csv` is an augmented dataset with latitude and longitude added for each site.


In [4]:
# Download and read the `spacex_launch_geo.csv`
from js import fetch
import io

URL = 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/datasets/spacex_launch_geo.csv'
resp = await fetch(URL)
spacex_csv_file = io.BytesIO((await resp.arrayBuffer()).to_py())
spacex_df=pd.read_csv(spacex_csv_file)

Now, you can take a look at what are the coordinates for each site.


In [5]:
# Select relevant sub-columns: `Launch Site`, `Lat(Latitude)`, `Long(Longitude)`, `class`
spacex_df = spacex_df[['Launch Site', 'Lat', 'Long', 'class']]
launch_sites_df = spacex_df.groupby(['Launch Site'], as_index=False).first()
launch_sites_df = launch_sites_df[['Launch Site', 'Lat', 'Long']]
launch_sites_df

Launch Site        Lat        Long
0   CCAFS LC-40  28.562302  -80.577356
1  CCAFS SLC-40  28.563197  -80.576820
2    KSC LC-39A  28.573255  -80.646895
3   VAFB SLC-4E  34.632834 -120.610745

Above coordinates are just plain numbers that can not give you any intuitive insights about where are those launch sites. If you are very good at geography, you can interpret those numbers directly in your mind. If not, that's fine too. Let's visualize those locations by pinning them on a map.


We first need to create a folium `Map` object, with an initial center location to be NASA Johnson Space Center at Houston, Texas.


In [6]:
# Start location is NASA Johnson Space Center
nasa_coordinate = [29.559684888503615, -95.0830971930759]
site_map = folium.Map(location=nasa_coordinate, zoom_start=10)

We could use `folium.Circle` to add a highlighted circle area with a text label on a specific coordinate. For example,


In [7]:
# Create a blue circle at NASA Johnson Space Center's coordinate with a popup label showing its name
circle = folium.Circle(nasa_coordinate, radius=1000, color='#d35400', fill=True).add_child(folium.Popup('NASA Johnson Space Center'))
# Create a blue circle at NASA Johnson Space Center's coordinate with a icon showing its name
marker = folium.map.Marker(
    nasa_coordinate,
    # Create an icon as a text label
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'NASA JSC',
        )
    )
site_map.add_child(circle)
site_map.add_child(marker)

and you should find a small yellow circle near the city of Houston and you can zoom-in to see a larger circle.


Now, let's add a circle for each launch site in data frame `launch_sites`


*TODO:*  Create and add `folium.Circle` and `folium.Marker` for each launch site on the site map


An example of folium.Circle:


`folium.Circle(coordinate, radius=1000, color='#000000', fill=True).add_child(folium.Popup(...))`


An example of folium.Marker:


`folium.map.Marker(coordinate, icon=DivIcon(icon_size=(20,20),icon_anchor=(0,0), html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'label', ))`


In [8]:
sites = launch_sites_df["Launch Site"]
sites

0     CCAFS LC-40
1    CCAFS SLC-40
2      KSC LC-39A
3     VAFB SLC-4E
Name: Launch Site, dtype: object

In [9]:
lat = launch_sites_df["Lat"]
lat

0    28.562302
1    28.563197
2    28.573255
3    34.632834
Name: Lat, dtype: float64

In [10]:
long = launch_sites_df["Long"]
long

0    -80.577356
1    -80.576820
2    -80.646895
3   -120.610745
Name: Long, dtype: float64

In [11]:
for i in range(0, len(sites)):
    print(sites[i])
    print(lat[i])
    print(long[i])

CCAFS LC-40
28.56230197
-80.57735648
CCAFS SLC-40
28.56319718
-80.57682003
KSC LC-39A
28.57325457
-80.64689529
VAFB SLC-4E
34.63283416
-120.6107455


In [12]:
# Initial the map
site_map = folium.Map(location= nasa_coordinate, zoom_start=4)
# For each launch site, add a Circle object based on its coordinate (Lat, Long) values. In addition, add Launch site name as a popup label
for i in range(0, len(sites)):
    circle = folium.Circle((lat[i], long[i]), radius=1000, color='#d35400', fill=True).add_child(folium.Popup(sites[i]))
    
    marker = folium.map.Marker(
            (lat[i], long[i]),
            # Create an icon as a text label
            icon=DivIcon(
                icon_size=(20,20),
                icon_anchor=(0,0),
                html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % sites[i],
                )
         )
    
    site_map.add_child(circle)
    site_map.add_child(marker)
    
site_map

The generated map with marked launch sites should look similar to the following:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_markers.png">
</center>


Now, you can explore the map by zoom-in/out the marked areas
, and try to answer the following questions:

*   Are all launch sites in proximity to the Equator line?
*   Are all launch sites in very close proximity to the coast?

Also please try to explain your findings.


## Task 2: Mark the success/failed launches for each site on the map


Next, let's try to enhance the map by adding the launch outcomes for each site, and see which sites have high success rates.
Recall that data frame spacex_df has detailed launch records, and the `class` column indicates if this launch was successful or not


In [13]:
spacex_df.tail(10)

Launch Site        Lat       Long  class
46    KSC LC-39A  28.573255 -80.646895      1
47    KSC LC-39A  28.573255 -80.646895      1
48    KSC LC-39A  28.573255 -80.646895      1
49  CCAFS SLC-40  28.563197 -80.576820      1
50  CCAFS SLC-40  28.563197 -80.576820      1
51  CCAFS SLC-40  28.563197 -80.576820      0
52  CCAFS SLC-40  28.563197 -80.576820      0
53  CCAFS SLC-40  28.563197 -80.576820      0
54  CCAFS SLC-40  28.563197 -80.576820      1
55  CCAFS SLC-40  28.563197 -80.576820      0

Next, let's create markers for all launch records.
If a launch was successful `(class=1)`, then we use a green marker and if a launch was failed, we use a red marker `(class=0)`


Note that a launch only happens in one of the four launch sites, which means many launch records will have the exact same coordinate. Marker clusters can be a good way to simplify a map containing many markers having the same coordinate.


Let's first create a `MarkerCluster` object


In [14]:
marker_cluster = MarkerCluster()


*TODO:* Create a new column in `launch_sites` dataframe called `marker_color` to store the marker colors based on the `class` value


In [15]:
launch_sites_df.head()

Launch Site        Lat        Long
0   CCAFS LC-40  28.562302  -80.577356
1  CCAFS SLC-40  28.563197  -80.576820
2    KSC LC-39A  28.573255  -80.646895
3   VAFB SLC-4E  34.632834 -120.610745

In [16]:
import numpy as np
spacex_df["marker_color"] = np.where(spacex_df["class"]==1, 'green', 'red')
spacex_df.head()

Launch Site        Lat       Long  class marker_color
0  CCAFS LC-40  28.562302 -80.577356      0          red
1  CCAFS LC-40  28.562302 -80.577356      0          red
2  CCAFS LC-40  28.562302 -80.577356      0          red
3  CCAFS LC-40  28.562302 -80.577356      0          red
4  CCAFS LC-40  28.562302 -80.577356      0          red

In [17]:
# Apply a function to check the value of `class` column
# If class=1, marker_color value will be green

In [18]:
# If class=0, marker_color value will be red
def get_marker_color(row):
    if row['class'] == 1:
        return 'green'
    else:
        return 'red'
# Apply the function to create the marker_color column
launch_sites_df['marker_color'] = spacex_df.apply(get_marker_color, axis = 1)

*TODO:* For each launch result in `spacex_df` data frame, add a `folium.Marker` to `marker_cluster`


In [19]:
# Add marker_cluster to current site_map
# Add marker_cluster to current site_map
site_map.add_child(marker_cluster)

# for each row in spacex_df data frame
# create a Marker object with its coordinate
# and customize the Marker's icon property to indicate if this launch was successed or failed, 
# e.g., icon=folium.Icon(color='white', icon_color=row['marker_color']
for index, record in spacex_df.iterrows():
    # TODO: Create and add a Marker cluster to the site map
    marker = folium.Marker(
            location = [record["Lat"], record["Long"]], 
            icon=folium.Icon(color='white', icon_color=record['marker_color']), 
            popup = record["Launch Site"],
        ). add_to(site_map)
        
    marker_cluster.add_child(marker)

site_map

Your updated map may look like the following screenshots:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_cluster.png">
</center>


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_cluster_zoomed.png">
</center>


From the color-labeled markers in marker clusters, you should be able to easily identify which launch sites have relatively high success rates.


## TASK 3: Calculate the distances between a launch site to its proximities


Next, we need to explore and analyze the proximities of launch sites.


Let's first add a `MousePosition` on the map to get coordinate for a mouse over a point on the map. As such, while you are exploring the map, you can easily find the coordinates of any points of interests (such as railway)


In [20]:
# Add Mouse Position to get the coordinate (Lat, Long) for a mouse over on the map
formatter = "function(num) {return L.Util.formatNum(num, 5);};"
mouse_position = MousePosition(
    position='topright',
    separator=' Long: ',
    empty_string='NaN',
    lng_first=False,
    num_digits=20,
    prefix='Lat:',
    lat_formatter=formatter,
    lng_formatter=formatter,
)

site_map.add_child(mouse_position)
site_map

Now zoom in to a launch site and explore its proximity to see if you can easily find any railway, highway, coastline, etc. Move your mouse to these points and mark down their coordinates (shown on the top-left) in order to the distance to the launch site.


Now zoom in to a launch site and explore its proximity to see if you can easily find any railway, highway, coastline, etc. Move your mouse to these points and mark down their coordinates (shown on the top-left) in order to the distance to the launch site.


In [21]:
from math import sin, cos, sqrt, atan2, radians

def calculate_distance(lat1, lon1, lat2, lon2):
    # approximate radius of earth in km
    R = 6373.0

    lat1 = radians(lat1)
    lon1 = radians(lon1)
    lat2 = radians(lat2)
    lon2 = radians(lon2)

    dlon = lon2 - lon1
    dlat = lat2 - lat1

    a = sin(dlat / 2)**2 + cos(lat1) * cos(lat2) * sin(dlon / 2)**2
    c = 2 * atan2(sqrt(a), sqrt(1 - a))

    distance = R * c
    return distance

*TODO:* Mark down a point on the closest coastline using MousePosition and calculate the distance between the coastline point and the launch site.

In [22]:
# find coordinate of the closet coastline
# distance_coastline = calculate_distance(launch_site_lat, launch_site_lon, coastline_lat, coastline_lon
launch_site_lat = 34.632834
launch_site_lon = -120.610745
coastline_lat = 34.63519
coastline_lon = - 120.62551
distance_coastline = calculate_distance(launch_site_lat, launch_site_lon, coastline_lat, coastline_lon)
distance_coastline

1.3764660599411633

In [23]:
# Create a marker for the launch site
launch_site_coordinate = [launch_site_lat, launch_site_lon]
launch_site_marker = folium.Marker(
    launch_site_coordinate,
    icon=folium.Icon(color='blue'),
)
launch_site_marker.add_to(site_map)

# Create a marker for the coastline with distance information
coastline_coordinate = [coastline_lat, coastline_lon]
coastline_marker = folium.Marker(
    coastline_coordinate,
    icon=DivIcon(
        icon_size=(20, 20),
        icon_anchor=(0, 0),
        html='<div style="font-size: 12px; color:#d35400;"><b>%s</b></div>' % "{:.2f} KM".format(distance_coastline),
    ),
)
coastline_marker.add_to(site_map)


*TODO:* Draw a `PolyLine` between a launch site to the selected coastline point


In [24]:
# Create a `folium.PolyLine` object using the coastline coordinates and launch site coordinate
# lines=folium.PolyLine(locations=coordinates, weight=1)
# Create a PolyLine to connect the launch site and the coastline
lines = folium.PolyLine(locations=[launch_site_coordinate, coastline_coordinate], weight=1)
lines.add_to(site_map)

# display the map
site_map

Your updated map with distance line should look like the following screenshot:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_distance.png">
</center>


*TODO:* Similarly, you can draw a line betwee a launch site to its closest city, railway, highway, etc. You need to use `MousePosition` to find the their coordinates on the map first


A railway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/railway.png">
</center>


A highway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/highway.png">
</center>


A city map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/city.png">
</center>


In [25]:
# Create a marker with distance to a closest city, railway, highway, etc.
# Draw a line between the marker to the launch site

In [26]:
# closest city
# distance_city = calculate_distance(launch_site_lat, launch_site_lon, city_lat, city_lon
launch_site_lat = 34.632834
launch_site_lon = -120.610745
city_lat = 34.63903
city_lon = - 120.45797
distance_city = calculate_distance(launch_site_lat, launch_site_lon, city_lat, city_lon)
distance_city

13.998589321648636

In [27]:
# Create a marker for the launch site
launch_site_coordinate = [launch_site_lat, launch_site_lon]
launch_site_marker = folium.Marker(
    launch_site_coordinate,
    icon=folium.Icon(color='blue'),
)
launch_site_marker.add_to(site_map)

# Create a marker for the city with distance information
city_coordinate = [city_lat, city_lon]
city_marker = folium.Marker(
    city_coordinate,
    icon=DivIcon(
        icon_size=(20, 20),
        icon_anchor=(0, 0),
        html='<div style="font-size: 12px; color:#d35400;"><b>%s</b></div>' % "{:.2f} KM".format(distance_city),
    ),
)
city_marker.add_to(site_map)


In [28]:
# Create a PolyLine to connect the launch site and the city
lines = folium.PolyLine(locations=[launch_site_coordinate, city_coordinate], weight=1)
lines.add_to(site_map)

# display the map
site_map

In [29]:
# closest railway
# distance_rail = calculate_distance(launch_site_lat, launch_site_lon, rail_lat, rail_lon
launch_site_lat = 34.632834
launch_site_lon = -120.610745
rail_lat = 34.63502
rail_lon = -120.62437
distance_rail = calculate_distance(launch_site_lat, launch_site_lon, rail_lat, rail_lon)
distance_rail

1.2704436850615237

In [30]:
# Create a marker for the launch site
launch_site_coordinate = [launch_site_lat, launch_site_lon]
launch_site_marker = folium.Marker(
    launch_site_coordinate,
    icon=folium.Icon(color='blue'),
)
launch_site_marker.add_to(site_map)

# Create a marker for the city with distance information
rail_coordinate = [rail_lat, rail_lon]
rail_marker = folium.Marker(
    rail_coordinate,
    icon=DivIcon(
        icon_size=(20, 20),
        icon_anchor=(0, 0),
        html='<div style="font-size: 12px; color:#d35400;"><b>%s</b></div>' % "{:.2f} KM".format(distance_rail),
    ),
)
rail_marker.add_to(site_map)

In [31]:
# Create a `folium.PolyLine` object using the railway coordinates and launch site coordinate
# lines=folium.PolyLine(locations=coordinates, weight=1)

lines = folium.PolyLine(locations=[launch_site_coordinate, rail_coordinate], weight=1)
lines.add_to(site_map)

# display the map
site_map

In [32]:
# closest highway
# distance_highway = calculate_distance(launch_site_lat, launch_site_lon, highway_lat, highway_lon)
launch_site_lat = 34.632834
launch_site_lon = -120.610745
highway_lat = 34.63473
highway_lon = -120.62345
distance_highway = calculate_distance(launch_site_lat, launch_site_lon, highway_lat, highway_lon)
distance_highway

1.1817325362821312

In [33]:
# Create and add a folium.Marker on your selected closest highway point on the map
# Display the distance between highway point and launch site using the icon property 
# Create a marker for the launch site
launch_site_coordinate = [launch_site_lat, launch_site_lon]
launch_site_marker = folium.Marker(
    launch_site_coordinate,
    icon=folium.Icon(color='blue'),
)
launch_site_marker.add_to(site_map)

# Create a marker for the highway with distance information
highway_coordinate = [highway_lat, highway_lon]
highway_marker = folium.Marker(
    highway_coordinate,
    icon=DivIcon(
        icon_size=(20, 20),
        icon_anchor=(0, 0),
        html='<div style="font-size: 12px; color:#d35400;"><b>%s</b></div>' % "{:.2f} KM".format(distance_highway),
    ),
)
highway_marker.add_to(site_map)

In [34]:
# Create a `folium.PolyLine` object using the highway coordinates and launch site coordinate
# lines=folium.PolyLine(locations=coordinates, weight=1)

lines = folium.PolyLine(locations=[launch_site_coordinate, highway_coordinate], weight=1)
lines.add_to(site_map)

# display the map
site_map

## FINDINGS

###### RAILWAYS

- The distance from VAFB SLC-4E launch site to the nearest railway (Santa Barbara Subdivision) is 1.27 km.
- It appears that the launch site is relatively close to a railway, which could be advantageous for transportation and logistics.

###### HIGHWAYS

- The distance from VAFB SLC-4E launch site to the nearest highway (Coast Road) is 1.18 km.
- The launch site is indeed close to a highway, which is beneficial for easy access and transportation of equipment.


###### COASTLINES

- The distance from VAFB SLC-4E launch site to the nearest coastline off Coast Road is 1.38 km.
- The launch site is also close to the coastline, which can be advantageous for launching rockets over the ocean, minimizing potential risks to populated areas.

###### CITIES

- The distance from VAFB SLC-4E launch site to the city of Lompoc is 14 km.
- The launch site is located at a significant distance from the city, which is a safety measure to avoid potential harm to populated areas in the event of an accident during a launch.

#### In summary, based on the calculated distances, the launch site appears to be strategically located with close proximity to railways, highways, and coastlines, while maintaining a safe distance from populated cities for safety reasons. These factors are important considerations for the logistics and safety of rocket launches.

After you plot distance lines to the proximities, you can answer the following questions easily:

*   Are launch sites in close proximity to railways?
*   Are launch sites in close proximity to highways?
*   Are launch sites in close proximity to coastline?
*   Do launch sites keep certain distance away from cities?

Also please try to explain your findings.


# Next Steps:

Now you have discovered many interesting insights related to the launch sites' location using folium, in a very interactive way. Next, you will need to build a dashboard using Ploty Dash on detailed launch records.


## Authors


[Pratiksha Verma](https://www.linkedin.com/in/pratiksha-verma-6487561b1/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork865-2022-01-01)


## Change Log


| Date (YYYY-MM-DD) | Version | Changed By      | Change Description      |
| ----------------- | ------- | -------------   | ----------------------- |
| 2022-11-09        | 1.0     | Pratiksha Verma | Converted initial version to Jupyterlite|


### <h3 align="center"> IBM Corporation 2022. All rights reserved. <h3/>
